Fix custom/pattern-userSpaceOnUse-userToBaseTransform.xhtml

rwlbuis opened the pull request:
#### c443c971c715d48e5b9031f62c0926d05f27310c
<pre>
Fix custom/pattern-userSpaceOnUse-userToBaseTransform.xhtml
<a href="https://bugs.webkit.org/show_bug.cgi?id=269234">https://bugs.webkit.org/show_bug.cgi?id=269234</a>

Reviewed by Nikolas Zimmermann.

When determining RenderSVGRoot as root layer we also calculate
any padding/border values as part of offsetFromAncestor, this
means for patterns painting would be wrongly offset for pattern
content. Instead, use the RenderSVGRoot, viewport as root layer
which discards any border/padding for the pattern content painting.

* LayoutTests/platform/glib/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-with-viewBox-expected.txt: Added.
* LayoutTests/platform/ios/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-with-viewBox-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-with-viewBox-expected.txt: Added.
* LayoutTests/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-with-viewBox-expected.png: Added.
* LayoutTests/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-with-viewBox-expected.txt: Added.
* LayoutTests/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-with-viewBox.xhtml: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintSVGResourceLayer):
(WebCore::RenderLayer::enclosingSVGRootLayer const): Deleted.
* Source/WebCore/rendering/RenderLayer.h:

Canonical link: <a href="https://commits.webkit.org/275117@main">https://commits.webkit.org/275117@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e5266794faa852eb95307c5d27f209689d43b5b0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/40900 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/19913 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/43278 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/43458 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/36991 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/43207 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22939 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/17244 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/33899 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/41474 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16867 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/35257 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/14522 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/14636 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/36247 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/44753 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/37121 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/36558 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/40306 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/15715 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12903 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/38675 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/17334 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/9190 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/17385 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/16978 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2840,7 +2840,12 @@ void RenderLayer::paintSVGResourceLayer(GraphicsContext& context, const AffineTr
     auto localPaintDirtyRect = LayoutRect::infiniteRect();
 
-    auto* rootPaintingLayer = enclosingSVGRootLayer();
+    auto* rootPaintingLayer = [this] () {
```
s/this/&/?

rwlbuis replied:
Done. But I think "this" was enough since the only thing needed from the context was calling parent(). Anyway, I don' think there is a big difference in this case.

nikolaszimmermann commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -2840,7 +2840,12 @@ void RenderLayer::paintSVGResourceLayer(GraphicsContext& context, const AffineTr
-    auto* rootPaintingLayer = enclosingSVGRootLayer();
+    auto* rootPaintingLayer = [this] () {
+        auto* curr = parent();
+        while (curr && !(curr->renderer().isAnonymous() && is<RenderSVGViewportContainer>(curr->renderer())))
```
Great, that this fixes the problem, without having to resort to context translations.

For completeness, we should duplicate the testcase that you fixed, and alter it to have a viewBox on the <svg> that effectively induces e.g. a scale(2) translation, to make sure paintSVGResourceLayer() is correct in these scenarios

rwlbuis replied:
I added the test, however I chose a smaller scale so all 3 columns still fit (I think r-w-t limits to 450x450 or so). Still need to see if platforms like gtk produce different txt/png results.

rwlbuis committed:
Fix custom/pattern-userSpaceOnUse-userToBaseTransform.xhtml
https://bugs.webkit.org/show_bug.cgi?id=269234

Reviewed by Nikolas Zimmermann.

When determining RenderSVGRoot as root layer we also calculate
any padding/border values as part of offsetFromAncestor, this
means for patterns painting would be wrongly offset for pattern
content. Instead, use the RenderSVGRoot, viewport as root layer
which discards any border/padding for the pattern content painting.

* LayoutTests/platform/glib/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-with-viewBox-expected.txt: Added.
* LayoutTests/platform/ios/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-with-viewBox-expected.txt: Added.
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/TestExpectations:
* LayoutTests/platform/mac-sonoma-wk2-lbse-text/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-with-viewBox-expected.txt: Added.
* LayoutTests/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-with-viewBox-expected.png: Added.
* LayoutTests/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-with-viewBox-expected.txt: Added.
* LayoutTests/svg/custom/pattern-userSpaceOnUse-userToBaseTransform-with-viewBox.xhtml: Added.
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::paintSVGResourceLayer):
(WebCore::RenderLayer::enclosingSVGRootLayer const): Deleted.
* Source/WebCore/rendering/RenderLayer.h:

Canonical link: https://commits.webkit.org/275117@main
